### ISSUE TITLE
Using the Triangle Inequality to Accelerate K-Means

### ISSUE THREAD

**Issue by krukah** [krukah/robopoker#34]
there are ways to eliminate redundant distance calculations in the kmeans clustering steps of the abstraction learning process. they exploit triangle inequalities of `Histogram` space to block out certain regions of the space from earth mover's distance (EMD) calculation. 

**it would be nice if the `Layer` data structure picked up a mutable field like `ExcludedHyperRectangle` that was updated online during EMD calculations.** 

this might pose a challenge for parallelism. it might not. depends on where in the call hierarchy the `updateExcludedHyperRectangle` would take place, and what its input arguments would be.

shoutout Elkan (2003)

https://cdn.aaai.org/ICML/2003/ICML03-022.pdf
https://dl.acm.org/doi/pdf/10.1145/312129.312248

**Comment by Drazcmd** [comment]
I've never done K-means clustering before ever (so this is all new to me!) but have been digging into the Elkan paper a bit in my free time. Sounds like the goal here would be skipping a bunch of the N*k calculations it's currently doing each time it calls [the "next()" method](https://github.com/krukah/robopoker/blob/8db10318b4ea9432e69e390af671805e4eeef0b0/src/clustering/layer.rs#L127)? Mainly using two 'bounds' vectors that track:
- lower bounds on distance from each point to each centroid (l(x, c) from the paper)
- upper bounds on distance from each point to its currently assigned centroid (u(x) from the paper)

, plus some math + a couple other temporary vectors?

Assuming that's correct I'll probably take a stab at this!
____

Side note: I _think_ we can still keep things well parallelized if we're careful - looking in the paper when talking about the accelerated algorithm I see
"""
When step (3) is implemented with nested loops, the outer loop can be over x or over c. For efficiency in Matlab and similar languages, the outer loop should be over c since k << n typically, and the the inner loop should be replaced by vectorized code that operates over all relevant x collectively
"""

**Comment by krukah** [comment]
this sounds very promising!! I agree with your summary of the approach. 

the parallelism point is interesting, assuming x is the "axis" of for each point in points, and c the "axis" of for each centroid of centroid? yeah looks like it could be a clean rayon injection if so.

**Comment by Drazcmd** [comment]
Currently WIP over in https://github.com/Drazcmd/robopoker/tree/feature/triangle-accel-kmeans. Been making good progress, though I still have a *lot* left to do (almost certainly won't work at all right now / I need to actually write some tests + benchmarks for it).


____


(In case anyone is curious about how I mapped it to the paper, currently I'm assuming the following:

- 'nearest neighbor' for a Point will effectively return the histogram corresponding to the currently-assigned centroid for a given point

- taking the average of points assigned to a center - i.e. to compute the new centroid histogram for the next training loop - can be done by choosing one point assigned to the center at the end of the loop and 'absorbing' all the histograms of the other points assigned to said centroid at that time

Hopefully those are both actually correct!)

**Comment by krukah** [comment]
this looks very promising!! don't worry about the cleanliness of the code, becomes much easier to consolidate and clean up after the fact. 

your analyses/assumptions are perfectly correct. it's also worth noting on the second point -- there is an edge case where a cluster ends up with 0 points assigned to it, and the type `Histogram` doesn't quite have a natural "zero/identity" element that we can default to. (it forms a semigroup, not a monoid)

**Comment by krukah** [comment]
also i've since made some organizational changes and renames that should be easy to reconcile with your branch, so i'm not too worried about resolving any potential merge conflicts since i've left the actual clustering logic alone. most of the recent changes i've pushed have been in the [mc]cfr module after all, so there should be strong isolation between our changes

**Comment by Drazcmd** [comment]
Finally getting around to running + testing the new code! Any suggestions on how you might approach verifying whether the outputs my new code produces will actually be "correct"? Were there any specific things / tests you did to verify the original code was working properly that might be useful?)

(Since even though it compiles, I suspect that it's very likely there's some bugs in it. I'd be pretty surprised if I did everything correctly on my first try!)

Current plans if not are:
- digging  into the other settings to see if I can further reduce the task even further (i.e. beyond just doing shortdeck)
- writing  a bunch of unit tests

____

Side note - normally I'd just do some comparisons of the `cargo run`, but currently I'm traveling and so using a (very) underpowered M1 8gb mac. Which means unfortunately:

- even with `--features shortdeck`, after an hour the old code had made 0% progress towards clustering TURN / hadn't gotten through any training iterations arias. (Also even just the _initialization_ took ~20 minutes).

- After 30m triangle-accelerated is also still at 0% / hasn't completed any iterations. That said I'm going to go ahead and leave it running overnight to see what happens. (e.g. I think I vaguely remember the paper mentioning the first runs could take a lot longer? not sure)

**Comment by Drazcmd** [comment]
Update: made some major progress! Once I fixed a couple dumb mistakes (e.g. extra clones, repeated distance calculations) the new algorithm started zooming. 

- Just with the low-hanging optimizations, short-deck turn clustering is now about **10x faster** than with the original algorithm. Specifically, on my machine ~30 minutes (vs ~5 hours).
    - TBD relative performance on the flop - running it right now, will update once I find out.
    - There's almost certainly plenty of ways we could optimize it further, but this is now actually narrowing in on the amount of time it takes just to _initialize_ all the clusters (i.e. ~13 minutes, i.e. only 2-3x the time). So this is probably 'good enough' for now.

- I _think_ the 3 resulting files in pgcopy/ are all indeed identical between the original code and the new triangle-accelerated code (i.e. as hoped). Going to go re-run with the original code one more time to confirm though.

Assuming everything goes well with the next couple runs, I'm probably going to go clean it up (+ add some tests) and then create a pull request.

**Comment by krukah** [comment]
you really cooked here, this is fantastic

- i'm just now reading update from 2w ago. i didn't internalize that results from the acceleration yielded identical results! it makes sense if i think about the geometry, but that's an incredible result for validation purposes. curious if there's even some test to write to assert this identity? might think about it later.
- i have to imagine that the speedup will work equivalently for flop, given the "black-box distance" property! 
- what i've done for testing identity between large pgcopy files is just sha256 them from the command line. it takes a few minutes but not prohibitively  long. excited to see results

**Comment by Drazcmd** [comment]
Makes sense! I'll take a stab at adding a test along those lines.

Still waiting to check for certain against the master branch, but here's some `sha256sum`s in the meantime **(shortdeck turn clustering)**:

Triangle Inequality accelerated: 
```
% sha256sum pgcopy/tiaccl_may30/*      
b4f86af7b5336c4809dcb6e6b0dc35199d1286f9310ab4e63585692fddbf2357  pgcopy/tiaccl_may30/isomorphism.turn
621573e1942d30fd23c1593dbff34c1a7df0785729cda0a9503bd5bed2b9b1dd  pgcopy/tiaccl_may30/metric.turn
732d5aa7cda41127f5a57aa679a038434ad3aa1a54fb953e65c2251e60628652  pgcopy/tiaccl_may30/transitions.turn
```

Unaccelerated:
```
% sha256sum pgcopy/unaccl_may30/*
b4f86af7b5336c4809dcb6e6b0dc35199d1286f9310ab4e63585692fddbf2357  pgcopy/unaccl_may30/isomorphism.turn
621573e1942d30fd23c1593dbff34c1a7df0785729cda0a9503bd5bed2b9b1dd  pgcopy/unaccl_may30/metric.turn
732d5aa7cda41127f5a57aa679a038434ad3aa1a54fb953e65c2251e60628652  pgcopy/unaccl_may30/transitions.turn
```
(So looks like we're probably good! Assuming I didn't accidentally modify or break the unaccelerated code when I was moving stuff around)

**Comment by krukah** [comment]
really exciting stuff, i looked through the code and its quite impressive . i like the separation of the algorithm into steps -- what do you think about creating a PR? i'd love to keep the implementation and style consistent, with respect to function naming, nesting, and abstraction levels

- i've looked through all your recent commit messages, appreciate the detail -- it makes it very easy to follow along
- in its final form, i'd like for the `TriIneq` implementation to be not only a first class citizen, but the default/only implementation
- wrt consistent function implementation details, i find control flow often simplifies when sticking to
  - only use imperative `for` when mutating local bindings; otherwise, use strict immutable refs and chain Iterator combinators
  - avoid multi-nesting, define smaller function blocks as necessary as multi nesting emerges

anyway i know this is perhaps still WIP since i see you still making changes and updating test cases, but this looks great and again, appreciate the documentation in commit messages!

**Comment by Drazcmd** [comment]
That all sounds good to me! I agree it still definitely needs some cleanup, was planning to do some other stuff (extracting it out to another file + giving it more granular/explicit inputs not locked against the values in street.rs) - making those changes you've suggested around the same time too shouldn't be too much extra work.

___

As far as making the PR right now, I'm actually not quite ready - turns out I was incorrect when I thought this was fully working in my most recent commit messages. And, actually I DID introduce a bug somewhere: in both the unaccelerated AND accelerated code the RMS values aren't going down on each iteration like they used to / are supposed to. Instead they're just hanging out at the value on the first iteration. (Which I think is also completely throwing off some of the time comparisons I did above)

The good news though is that since it's happening in both, it means we can _probably_ narrow it down to my changes to this block of code:
```
          let ref mut next = self.next();
          let ref mut last = self.kmeans;
          std::mem::swap(next, last);
          progress.inc(1);
```
- so hopefully shouldn't take too much more time to resolve. (That said I'm going to need to do a couple more runs to confirm!)

**Comment by Drazcmd** [comment]
E.g. comparing (both unaccelerated) code before and after my changes:

```
17:17:26 [INFO] clustering  kmeans              turn
17:25:06 [DEBUG] abstraction cluster RMS error   0.029961692                     
17:36:44 [DEBUG] abstraction cluster RMS error   0.027634056                     
17:48:21 [DEBUG] abstraction cluster RMS error   0.026905635   

vs

05:21:44 [INFO] clustering kmeans (unaccelerated)turn
05:29:23 [DEBUG] abstraction cluster RMS error   0.029961692                     
05:37:04 [DEBUG] abstraction cluster RMS error   0.029961692                     
05:44:43 [DEBUG] abstraction cluster RMS error   0.029961692    
```

and then

```
19:10:56 [INFO] clustering  kmeans              flop
21:11:06 [DEBUG] abstraction cluster RMS error   0.043196086                     
17:25:11 [DEBUG] abstraction cluster RMS error   0.042680442
16:23:40 [DEBUG] abstraction cluster RMS error   0.041885197                     

vs

10:39:36 [INFO] clustering kmeans (unaccelerated)flop
12:49:22 [DEBUG] abstraction cluster RMS error   0.048136037                     
14:59:10 [DEBUG] abstraction cluster RMS error   0.048136037                     
17:09:09 [DEBUG] abstraction cluster RMS error   0.048136037    
```
(same behavior, I think numbers differ probably only because the turn itself is basically unclustered in the latter. And that if we were to reuse the original unaccelerated code's output it would be doing `0.043196086 0.043196086 0.043196086`)

**Comment by Drazcmd** [comment]
Some good news: I've finally refactored enough to add proper unit tests! Going forwards we should be able to detect the following if they happen again:
- accelerated and original algorithm 'disagreeing' at any step (from RMS not matching)
- accelerated algorithm RMS not decreasing
- original algorithm RMS not decreasing

Some less-good news: I'm likely going to be pretty busy with work for the next couple weeks/months, so I probably won't be able to devote as much time to this. That said I'll at least try to finish up the refactoring + send out the PR this weekend (and hopefully close the issue here) before that all happens :)

**Comment by krukah** [comment]
i'm excited to review and merge all of this, and just wanted to say thanks for your contribution! it all really looks very high quality and well-thought out, and most importantly massive performance boost unlock. i'll take some time reviewing it, although at first glance i like a lot of the design choices i see you've made, so looking forward to it!
